tests/csapi/apidoc_room_event_test.go

```diff
@@ -0,0 +1,206 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+)
+
+func stringSliceContains(slice []string, s string) bool {
+	for _, value := range slice {
+		if value == s {
+			return true
+		}
+	}
+	return false
+}
+
+// sytest: Events come down the correct room
+func TestEventsInCorrectRoom(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+	const (
+		roomAmount     = 10
+		goConcurrency  = 4
+		channelTimeout = 10 * time.Second
+	)
+
+	// Create all rooms
+
+	var doneRooms = make(chan string, roomAmount)
+	var todoRooms = make(chan bool, goConcurrency)
+
+	for i := 0; i < goConcurrency; i++ {
+		go func() {
+			for {
+				_, ok := <-todoRooms
+
+				if !ok {
+					return
+				}
+
+				roomId := alice.CreateRoom(t, map[string]interface{}{
+					"preset": "public_chat",
+				})
+
+				doneRooms <- roomId
+			}
+		}()
+	}
+
+	for i := 0; i < roomAmount; i++ {
+		todoRooms <- true
+	}
+
+	close(todoRooms)
+
+	var rooms []string
+
+	for i := 0; i < roomAmount; i++ {
+		select {
+		case newRoom := <-doneRooms:
+			rooms = append(rooms, newRoom)
+		case <-time.After(channelTimeout):
+			t.Fatalf("Room creation timed out, got to %d rooms (from %d)", i, roomAmount)
+		}
+	}
+
+	// Send events to all rooms with corresponding roomID
+
+	// get current next_batch
+	_, since := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+	type txnAndId struct {
+		txn    int
+		roomId string
+	}
+
+	var doneEvents = make(chan string, roomAmount)
+	var todoEvents = make(chan txnAndId, goConcurrency)
```

Comment:
It doesn't make sense to me to set the buffer of this channel to `goConcurrency` and not `roomAmount`. `todoEvents` is injected with `roomAmount` items, so this will cause things to block on sending (because `goConcurrency < roomAmount`) for no justified reason.

Reply:
I'm fairly sure it has little effect, yes, the only effect i can think of is that it will block until `roomAmount - goConcurrency` items have been processed (in `doneEvents`), through concurrency, so it makes the below bit where it awaits timeout a bit of a dud unless the last `goConcurrency` amount of items take that long.

In other words, changing this to `roomAmount` will make it detect timeouts earlier, but I'll apply the suggestion regardless. 👍